File: 0051-n-queens/solution.py
class Solution:
    
    def solveNQueens(self, n: int) -> List[List[str]]:
        
        if n == 1:
            return [["Q"]]
        
        self.n         = n
        self.result    = []
        self.numQueens = 0
       
        def canAddSafely(r, c):
            diag = r + c
            antiDiag = r - c
            
            safe = (c not in self.cols)
            safe = safe and (diag not in self.diags)
            safe = safe and (antiDiag not in self.antiDiags)
            
            return safe
        
        def addQueen(r, c):
            self.board[r][c] = 'Q'
            self.numQueens += 1

            self.antiDiags.add(r-c)
            self.diags.add(r+c)
            self.cols.add(c)
        
        def deleteQueen(r, c):
            self.board[r][c] = '.'
            self.numQueens -= 1

            self.antiDiags.remove(r-c)
            self.diags.remove(r+c)
            self.cols.remove(c)
        
        def printBoard():
            for row in self.board:
                print(row)
            print("\n\n")
        
        def solve():
            for c in range(self.n):
                if canAddSafely(self.row, c):
                    addQueen(self.row, c)
                    
                    if self.row == self.n - 1 and self.numQueens == self.n:
                        self.result.append(["".join(row) for row in copy.deepcopy(self.board)])
                    
                    elif self.row != self.n-1:
                        self.row += 1
                        solve()
                        self.row -= 1
                    
                    deleteQueen(self.row, c)
                    
        
        for i in range(self.n):
            self.cols, self.diags, self.antiDiags = set(), set(), set()        
            self.board = [["."]*self.n for _ in range(self.n)]
            self.numQueens = 0
            self.row = 0
            
            addQueen(0, i)
            
            self.row = 1
            
            solve()
            
            
        return self.result
        
        

File: 0051-n-queens/solution.cpp
class Solution 
{
private:
    int n, row, numQueens;
    unordered_set<int> cols, diags, antiDiags;
    vector<vector<string>> result;
    vector<string> board;
public:
    void getEmptyBoard()
    {
        board = vector<string>();
        for(int r = 0; r < n; r++)
        {
            string boardRow = "";
            for(int c = 0; c < n; c++)
            {
                boardRow += ".";
            }
            board.push_back(boardRow);
        }
    }
    
    void addQueen(int r, int c)
    {
        int diag = r + c;
        int antiDiag = r - c;
        
        cols.insert(c);
        diags.insert(diag);
        antiDiags.insert(antiDiag);
        
        board[r][c] = 'Q';
        numQueens++;
    }
    
    void removeQueen(int r, int c)
    {
        int diag = r + c;
        int antiDiag = r - c;
        
        cols.erase(c);
        diags.erase(diag);
        antiDiags.erase(antiDiag);
        
        board[r][c] = '.';
        
        numQueens--;
    }
    
    bool canSafelyAdd(int r, int c)
    {
        int diag = r + c;
        int antiDiag = r - c;
        
        if(diags.find(diag) != diags.end())
            return false;
        if(antiDiags.find(antiDiag) != antiDiags.end())
            return false;
        if(cols.find(c) != cols.end())
            return false;
        
        return true;
    }
    
    void printBoard()
    {
        for(auto & row: board)
        {
            cout<<row<<endl;
        }
        
        cout<<endl; cout<<endl;
    }
    
    void solve()
    {
        for(int c = 0; c < n; c++)
        {
            // cout<<"row "<<row<<" col "<<c<<endl;
            if(canSafelyAdd(row, c))
            {
                // cout<<"can safely add"<<endl;
                addQueen(row, c);            
                
                // printBoard();
                
                if(row == n-1)
                {
                    if(numQueens == n)
                    {
                        vector<string> copy;
                        for(auto & row: board)
                        {
                            copy.push_back(row);
                        }
                        
                        result.push_back(copy);
                    }
                }
                else if(row < n-1)
                {
                    row++;
                    solve();
                    row--;
                }
                
                removeQueen(row, c);
            }
        }
    }
    
    
    vector<vector<string>> solveNQueens(int n) 
    {
        if(n == 1)
        {
            return {{"Q"}};
        }
        numQueens = 0;
        this->n = n;
        
        
        for(int i = 0; i < n; i++)
        {
            cols.empty(); diags.empty(); antiDiags.empty();
            
            getEmptyBoard();
            row = 0;
            
            addQueen(0, i);

            row = 1;
            solve();
            
            removeQueen(0, i);

        
        }
        
        return result;
    }
};
